### DISCUSSION TITLE
How to send/receive messages to the server from python environment

### DISCUSSION THREAD

**Discussion by bellatrics** [evennia/evennia#2800] [Evennia Questions] [answered]
Hi everyone,

My project is about creating a game and then optimizing/solving it with some reinforcement learning techniques. For this to happen, I need to be able to send/receive messages from python environment. Does anyone have some code snippets for doing this? I would much appreciate it.

I was looking into autobahn/twisted [WebSocketClientFactory](https://autobahn.readthedocs.io/en/latest/reference/autobahn.asyncio.html#autobahn.asyncio.websocket.WebSocketClientFactory) it looks like a valid approach, but I haven't succeeded yet.

Thanks for your help.

**Comment by InspectorCaracal** [comment]
I'm not entirely sure what you're trying to do, but the game has a `py` command which allows you to open a python shell inside the game client and you could send lines of python to the server that way?

[Here is the documentation](https://www.evennia.com/docs/latest/api/evennia.commands.default.system.html?highlight=py#evennia.commands.default.system.CmdPy)

This line explains the functionality which I think might suit your needs:
> Without argument, open a Python console in-game. This is a full console, accepting multi-line Python code for testing and debugging. Type exit() to return to the game. If Evennia is reloaded, the console will be closed.

Using it in the game looks like this:
```py
Evennia Interactive Python mode
Python 3.9.2 (default, Feb 28 2021, 17:03:44) 
[GCC 10.2.1 20210110] on linux
[py mode - quit() to exit]
>>> test = "  a test string  "
[py mode - quit() to exit]
>>> print(test)
  a test string  
[py mode - quit() to exit]
>>> print(test.strip())
a test string
[py mode - quit() to exit]
```

**Comment by bellatrics** [comment]
Thanks for the fast reply. I think I didn't express myself clear. Let me try again.

I have "mygame" running on localhost

What I need is being able to do the following...


Python 3.9.2 (default, Feb 28 2021, 17:03:44) 

class EvenniaClient(X):
      """"""

>> client = EvenniaClient("localhost", 400x)
>> client.make_action("go room x")
>> client.read_last_reply()
"you are in a room X"
>> client.make_action("pick pop-corn")
>> client.read_last_reply()
"...."

So I have an interface with the server. I tried using telnet client, but didn't succeeded. 
I hope this makes it more understandable.

Thanks again.

**Reply by InspectorCaracal** [reply]
I can see that you're trying to make your own client that connects to Evennia, but I don't understand why you need to do that instead of using any other client. You can connect to localhost with them, or on your browser for the webclient.

**Reply by bellatrics** [reply]
I think you just assumed that I am a evennia expert :) 
I just wanted to do things programmatically, but not "with my browser"
I found a code example (will post it lower)

**Comment by Antrare** [comment]
Hi, based on my understanding of what you're looking for, I think it's
essentially a telnet proxy. You telnet into your ML environment which then
connects to a mud and performs its tasks.  Try googling "telnet proxy
python" here is one of the results which appears to meet your needs.

https://gist.github.com/night-crawler/6213578

I built one a long time ago so I could be competitive in a mud where
latency was a killer... and I'm in Australia with minimum 200ms to the US.
So I put my proxy in the same area as the mud and added triggers/etc. there
rather than my telnet client.

Good luck on your project!

**Comment by Griatch** [comment, marked as answer]
You can talk to the live server over 
- telnet (the `dummyrunner` in `evennia/server/profiling/dummyrunner.py` is an example of doing this) 
- webclient (the main webclient is an example of how you do this - you can investigate its javascript code and adopt that or you could make your own simple websocket asyncio client that communicates using JSON data whose form is outlined in the Evennia communication spec set up by `evennia/web/static/webclient/js/evennia.js`). 
- You can't send raw Python to the server unless you are connected and logged in and are using the `py` command (which still means it technically communicates using telnet or websocket).

**Comment by bellatrics** [comment]
Thanks everyone for suggestions. I will explore the links.

I am posting a small code snippet that can do the thing:

>import telnetlib
>
>tn = telnetlib.Telnet("localhost", 4000)
>tn.write(b"connect user password\n")
>s = tn.read_very_eager().decode('ascii')
>print(s)
>tn.write(b"inv\n")
>s = tn.read_eager().decode('ascii')
>print(s)

**Reply by Griatch** [reply]
Yep :) The dummyrunner link I gave is basically doing something like this, but using twisted instead.
